### PR TITLE
chore: upgrade viem to v2

### PR DESCRIPTION
**Pull request by darrylyeo**
## Why is this change needed?

* Upgrades [`viem`](https://viem.sh/) from v1 to [v2](https://viem.sh/docs/migration-guide#2xx-breaking-changes). More context in [this cast](https://warpcast.com/v/0xffa5abfb).
  * [Maximizes tree-shaking](https://viem.sh/docs/clients/custom#tree-shaking) by importing action functions.
  * Adds type parameters to helper wrapper classes to ensure inputs align with `viem`'s deeply-nested conditional types.

## Merge Checklist

_Choose all relevant options below by adding an `x` now or at any time before submitting for review_

- [ ] PR title adheres to the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) standard
- [ ] PR has a [changeset](https://github.com/farcasterxyz/hub-monorepo/blob/main/CONTRIBUTING.md#35-adding-changesets)
- [ ] PR has been tagged with a change label(s) (i.e. documentation, feature, bugfix, or chore)
- [ ] PR includes [documentation](https://github.com/farcasterxyz/hub-monorepo/blob/main/CONTRIBUTING.md#32-writing-docs) if necessary.

---

## PR-Codex overview
This PR upgrades `viem` to version 2 across multiple packages. It also introduces new actions like `getChainId` and `watchBlockNumber`.

### Detailed summary
- Upgraded `viem` to version 2 in various packages
- Added new actions like `getChainId` and `watchBlockNumber`
- Updated usage of `viem` functions in different files

> The following files were skipped due to too many changes: `packages/hub-nodejs/examples/hello-world/index.ts`, `apps/hubble/src/eth/l2EventsProvider.test.ts`, `packages/hub-nodejs/examples/contract-signatures/index.ts`, `apps/hubble/src/eth/l2EventsProvider.ts`, `yarn.lock`, `packages/hub-nodejs/examples/contract-signatures/yarn.lock`, `packages/hub-nodejs/examples/hello-world/yarn.lock`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: packages/hub-nodejs/examples/contract-signatures/package.json [hunk 1]
```diff
@@ -11,11 +11,11 @@
...
-    "viem": "^1.19.13"
+    "viem": "^2.17.4"
   },
   "devDependencies": {
     "hardhat": "^2.19.2",
     "tsx": "^3.12.5",
     "typescript": "^5.0.0"
   }
-}
+}
```
Comment (dalechyn) [inline packages/hub-nodejs/examples/contract-signatures/package.json:21]: Needs a newline?

**Review by dalechyn — COMMENTED** [review]
Why did all of the `<clientInstance>.<instanceActionMethod>(...)` changed to `<globalActionMethod>(<clientInstance>, ...)`?

I.e: `testClient.mine(...)` changed to `mine(testClient)`.

AFAIK those methods are still accessible.
What's the motivation behind this?

**Comment by darrylyeo** [comment]
> Why did all of the `<clientInstance>.<instanceActionMethod>(...)` changed to `<globalActionMethod>(<clientInstance>, ...)`?

I chose to switch to this pattern because:

1. it allows for tree-shaking (see [docs](https://viem.sh/docs/clients/custom#tree-shaking)).
2. it makes it easier to derive types for helper functions using e.g. `Parameters<>` rather than chasing down the exact helper type + params used internally by the library.

**Comment by varunsrin** [comment]
@darrylyeo we have some lints failing, which are blocking the tests from running

File: apps/hubble/src/eth/l2EventsProvider.test.ts [hunk 2]
```diff
@@ -86,6 +87,7 @@ describe("process events", () => {
 
     l2EventsProvider = new L2EventsProvider(
       hub,
+      // @ts-ignore
```
Comment (varunsrin) [inline apps/hubble/src/eth/l2EventsProvider.test.ts:90]: @darrylyeo can you add more context about why the ignore is necessary?

File: apps/hubble/src/eth/l2EventsProvider.ts [hunk 3]
```diff
@@ -174,6 +183,7 @@ export class L2EventsProvider {
 
     const provider = new L2EventsProvider(
       hub,
+      // @ts-ignore
```
Comment (varunsrin) [inline apps/hubble/src/eth/l2EventsProvider.ts:186]: why is an ignore necessary here?

File: apps/hubble/src/eth/l2EventsProvider.ts [hunk 4]
```diff
@@ -632,6 +642,7 @@ export class L2EventsProvider {
       {
         address: storageRegistryAddress,
         abi: StorageRegistry.abi,
+        // @ts-ignore
```
Comment (varunsrin) [inline apps/hubble/src/eth/l2EventsProvider.ts:645]: Is the ignore necessary here?

File: .changeset/violet-steaks-rule.md [hunk 5]
```diff
@@ -0,0 +1,7 @@
+---
+"@farcaster/hub-nodejs": minor
```
Comment (varunsrin) [inline .changeset/violet-steaks-rule.md:2]: should be a patch version. 

minor versions are for breaking upgrades in farcaster's versioning system

**Review by varunsrin — CHANGES_REQUESTED** [review]

**Comment by darrylyeo** [comment]
Okay, managed to find the right type parameters to satisfy the underlying viem actions. Had a lot of trouble earlier getting viem's conditional return types to match up as TypeScript was unable to infer the parameters from the loosely-typed helper methods.

File: .changeset/violet-steaks-rule.md (hunk 5 above)
Comment (Wazzymandias) [reply .changeset/violet-steaks-rule.md]: also need to run `yarn lint` to format

File: apps/hubble/src/eth/l2EventsProvider.ts [hunk 6]
```diff
@@ -807,11 +868,16 @@ export class L2EventsProvider {
   private async detectFilterSupport() {
     // Set up a client with fewer retries and shorter timeout
     const urls: string[] = [];
-    this._publicClient.transport["transports"].forEach((transport) => {
-      if (transport?.value) {
-        urls.push(transport.value["url"]);
-      }
-    });
+
+    if (this._publicClient.transport["transports"]?.[0]?.value?.["url"]) {
```
Comment (CassOnMars) [inline apps/hubble/src/eth/l2EventsProvider.ts:872]: why did this check get added prior to iterating?

Comment (darrylyeo) [reply apps/hubble/src/eth/l2EventsProvider.ts:872]: It was an edge case caught by the type system – depending on how this class is initialized, the transport object is not necessarily guaranteed to be an [`HttpTransport`](https://viem.sh/docs/clients/transports/http.html).

**Comment by darrylyeo** [comment]
For anyone curious, I wrote about my experience implementing TypeScript type parameters here:
https://warpcast.com/darrylyeo/0x01dbdbaa

**Pull request merged**